### ISSUE TITLE
Backgrounded jobs

### ISSUE THREAD

**Issue by rpdelaney** [twolfson/sexy-bash-prompt#80]
I background things a lot. Especially with the editor. Then I come back to the window in tmux and forgot if I have the job backgrounded and type `fg` to check, which wastes time.

I thought the prompt could report this for me. So I threw something together: 

https://github.com/rpdelaney/sexy-bash-prompt/tree/jobs

I guess this might be scope creep. I'm happy to leave it in my own fork.

But it's easy and useful. Shell builtins mostly. What do you think?

**Comment by twolfson** [comment]
I don't think this is unreasonable, though it feels off as built-in behavior. The symbol is confusing (though I guess most of ours are) and this portion is very unrelated to `git`

I've also been wanting to add long-running command timer for a while but never put in the effort for it (e.g. show `(2:33)$` when last command took 2 minutes and 33 seconds)

Maybe there's a good middleground for documenting "plugins" like this in our README for people to add as they see fit:

- Allow defining function `sexy_bash_prompt_after_git_progress`
- Allow defining function `sexy_bash_prompt_before_symbol`

**Comment by rpdelaney** [comment]
> The symbol is confusing 

Yeah. I just picked something random; I'm not attached to it. I didn't put much thought into it. If this were to be in master, we could think of something better.

> this portion is very unrelated to git

Well, we _are_ "sexy-bash-prompt" not "sexy-git-prompt". But I agree the prompt is very git focused and so this could be seen as scope creep.

> I've also been wanting to add long-running command timer for a while but never put in the effort for it (e.g. show (2:33)$ when last command took 2 minutes and 33 seconds)

I hadn't thought of that. Another feature I might like is some kind of indication of whether the last command exited nonzero though.

> Maybe there's a good middleground for documenting "plugins" like this

This might be the way to go. I would name the functions differently, but if you think it's a good idea in general, I could whip something up. I would need help with the tests though -- even after all this time I still haven't grok'd how they work. heh.

**Comment by twolfson** [comment]
Hmmm, yea. Now I'm on the other side of the pendulum. I feel like attempting to create a plugin framework is adding an unnecessary layer of abstraction. Maybe we could include code snippets in the `README` (unsure if too bulky) or an `examples` folder (e.g. `examples/stopped-tasks`)

Also, yes I like the non-zero exit idea as well. Always hate second guessing myself and doing a `echo $?`

**Comment by rpdelaney** [comment]
> Maybe we could include code snippets in the `README` (unsure if too bulky)

I dunno how you are visualizing this but it would be very bulky, I think. It might be better just to have forks or branches. That's extra maintenance overhead, but the master branch is so stable at this point, it might not be a big deal. 

What caused you to change your mind? Supporting something like a `before_sbp()` and a `after_sbp()` function (again, with better names) seems like much less bulk than that and it gives more flexibility to the user. Otherwise we are now forced to make decisions like "Does _everyone_ need the prompt to tell them how many backgrounded jobs they have?" Do we need to build in flags to disable some options? At what point does the prompt become too busy? etc

> Also, yes I like the non-zero exit idea as well.

Cool. Very easy to do once we figure out how. :-) I can open a separate issue to discuss that.

**Comment by twolfson** [comment]
When I think of the extreme version of customizaton, I imagine needing a function for every section between each dialogue (e.g. someone wants to put a timestamp after hostname)

For what it's worth, I already see 4 places that could be common (not even including augmenting something like the `git` info to support other VCSs) (i.e. before `git` branch, before `git` symbol, after `git` symbol, before prompt symbol)

This then feels like we're adding unnecessary abstraction to allow injecting at any point in the script. In reality, it'd be better performance and easier maintenance for everyone if people simply forked for what they needed OR we have explicit supported opt-in/opt-out features (so we don't need to vaguely guess where content goes) (you mentioned this in your last comment)

I know I'm jumping around a lot on this, that's because it feels like a small change which has huge implications for future revisions and I've over-engineered major revisions in the past

I'm going to take some more time to think about this

**Comment by rpdelaney** [comment]
Hm. Sounds like a slippery slope fallacy honestly. If supporting injection in between every token is too much to maintain, then we should simply not support that. There's nothing requiring us to go to such lengths until we find a reason to. And we probably won't.

I mean, rather than do that, it would be better if we would rebuild the whole thing from scratch so that the prompt itself is a collection of tokens and sexy-bash-prompt is just a suggested or default configuration for it. That is wholly out of scope for this issue.

**Comment by twolfson** [comment]
Yea, it's definitely a slipper slope mentality. Unsure it's that far of a
leap such that it's a fallacy here though. We've gotten plenty of one off
customization requests over the years

I might be missing something but isn't our prompt already pretty modular?
The PS1 is rather broken up

**Comment by rpdelaney** [comment]
>  Unsure it's that far of a leap such that it's a fallacy here though. We've gotten plenty of one off customization requests over the years

I mean, we get the requests, but we decide where the line gets drawn. So we decide how slippery the slope is.

> I might be missing something but isn't our prompt already pretty modular? The PS1 is rather broken up

I dunno if I would call it modular. But please please don't misunderstand me. I am _not_ suggesting re-engineering it like that. I'm saying if we _were_ to accommodate every 1-off customization request (which I don't think we should), it would be easier to rebuild it in a fully modular way, than to just keep building stuff into it.

Anyway. Where do we go from here? I guess the options are:

1. Do nothing: 
* Good, because we don't have to do anything
* Good, because there's no scope creep. 
* Bad, because the feature might be useful to someone (although they can always fork it like I have).
2. Flesh out my branch, polish it up, and merge it into master: 
* Basically the opposite of No.1
3. Add a "before_prompt()" and "after_prompt()" of some kind: 
* Good, because users can customize their prompts with whatever weird stuff they want to add without bugging us to restructure everything
* Bad, because of the threat of the slippery slope that we wind up building a whole scaffolding tokenizer structure over time
4. Rebuild the whole prompt to be nothing but modules (something like `${TOKEN1_COLOR}prompt_token1() ${TOKEN2_COLOR}prompt_token2()...`
* Good, because then the customization issue is dead forever since anyone can do whatever they want on their own
* Bad, because it's a lot of work, basically scrapping everything in here and starting over from scratch
* Bad, because it's a cartoonish instance of scope creep

I don't have super strong opinions except that I don't like No. 4. If that is done it should be a completely separate project. And I don't want to do it. Other than that, I lean ever so slightly toward No 3. What do you think?

**Comment by twolfson** [comment]
We're missing opt-in/opt-out flags in this list but maybe that's option 2

I'm still going to take some time to about this more rather than making a
decision immediatly

**Comment by rpdelaney** [comment]
Sure, take your time. I'm happy to have my prompt in my fork.

**Comment by twolfson** [comment]
Coming back to this finally. I do think opt-in flags are the ideal extension structure for this repo -- however, when I think about them in application to this specific issue (background jobs), I feel like they're not applicable

I think it's that I don't see it a widespread use case (I think a minority of my coworkers have used stopped jobs in the past)

When I move the opt-in flag scope to things like error code detection and long running task timers, I feel like they are so useful across all programmers that it's silly to build out such an opt-in system

To summarize, let's keep backgrounded jobs as a forkable issue (maybe we can add links to other repos in the README as a middleground) and open issues for the 2 enhancements brought up (long running task timer, error code detection)

Thoughts?

**Comment by rpdelaney** [comment]
Yeah that's fine. I'll make separate issues for opt-in flags, and for tracking exit status. Perhaps you should do the same for timers since I don't understand the use case as well there.

**Comment by twolfson** [comment]
:+1: Will do
